lib/spack/spack/util/compression.py

```diff
@@ -48,30 +188,94 @@ def _unzip(archive_file):
-        with ZipFile(archive_file, 'r') as zf:
-            zf.extractall(destination_abspath)
+        zf = ZipFile(archive_file, 'r')
+        zf.extractall(destination_abspath)
+        zf.close()
     except ImportError:
         unzip = which('unzip', required=True)
         unzip.add_default_arg('-q')
-        return unzip
+        unzip(archive_file)
+    return destination_abspath
+
 
+def composer(funcA):
+    """Utility method currying function pointers
+    returns a function pointer to be called with
+    a function to be curried with the current function
+    argument.
 
-def decompressor_for(path, extension=None):
+    Args:
+        funcA : Function to be curried.
+    """
+    def b(funcB):
+        def c(*args, **kwargs):
+            return funcA(funcB(*args, **kwargs))
+        return c
+    return b
+
+
+def decompressor_for(path, ext=None):
+    """Wrapper for select_decompressor_for, returns
+    a function pointer to appropriate decompression
+    algorithm.
+
+    On Unix - simply invokes select_decompressor_for
+
+    On Windows - archives with one extension are
+    passed through to select_decompressor_for. Multiple
+    extension archives are decomposed into their component
+    extensions and the requsite decompression algorithms are
+    curried and returned as a single callable function pointer.
+
+    Args:
+        path (str): path of the archive file requiring decompression
+        ext (str): Extension of archive file
+    """
+    if ext:
+        if not allowed_archive(ext):
+            raise CommandNotFoundError("Unrecognized file extension")
+    if sys.platform == 'win32':
+        if ext is None:
+            ext = extension(path)
+        ext_l = ext.split(".")
+        # special case as there is no consistently
+        # available native python tool for .Z files
+        if not ext_l[1:] or re.search(r'.Z|xz', ext):
+            return select_decompressor_for(path, ext)
+        else:
+            return composer(
+                decompressor_for(path, ext_l[0])
+            )(
+                decompressor_for(
+                    path, ext=".".join(ext_l[1:]))
+            )
+    else:
+        return select_decompressor_for(path, ext)
+
+
+def select_decompressor_for(path, extension=None):
     """Get the appropriate decompressor for a path."""
     if ((extension and re.match(r'\.?zip$', extension)) or
             path.endswith('.zip')):
         return _unzip
     if extension and re.match(r'gz', extension):
         return _gunzip
     if extension and re.match(r'bz2', extension):
-        bunzip2 = which('bunzip2', required=True)
-        return bunzip2
-    tar = which('tar', required=True)
-    tar.add_default_arg('-oxf')
-    return tar
+        return _bunzip2
+    # Catch unexpected extensions and .Z files
+    # here. Python does not have native support
+    # of any kind for .Z files. In these cases, fall back
+    # to system defined strategy.
```

Comment:
I've been thinking more about the basis for using a config file as a fallback and it seems somewhat unlike our general logic here: if there is a `.Z` file, then we should probably be constructing an `Executable()` from `which(7z)` in this module, vs. falling back to the config.

BTW, do any of the tests confirm the success of `7z`? I assume it would need arguments (e.g. like `tar` needs `-x` to know that it should extract).

I think that would also address another concern I have with this: the fallback config is a list of possible strategies, which right now includes tar and 7z; therefore, it seems like this might return `tar` for a `.Z` file.